### PR TITLE
Implement some missing "IsInside"

### PR DESCRIPTION
**Pull request by couet**
As said in [this issue](https://github.com/root-project/root/issues/14453) TEllipse was missing the IsInside method.
Some other classes were missing it too (TCrown) or some were wrong (TDiamond). This PR implements the missing ones and fixes the wrong ones.

Also, a new graphics example `inside.C` has been added.

### CODE REVIEW

File: graf2d/graf/inc/TCrown.h [hunk 1]
```diff
@@ -25,11 +25,12 @@ class TCrown : public TEllipse {
    TCrown(const TCrown &crown);
    ~TCrown() override;
 
-   void   Copy(TObject &crown) const override;
+   void    Copy(TObject &crown) const override;
```
Comment (vepadulano) [inline graf2d/graf/inc/TCrown.h:28]: Spurious extra space

**Review by vepadulano — CHANGES_REQUESTED** [review]

File: tutorials/graphics/inside.C [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+/// \macro_code
+///
+/// \author Olivier Couet
+
+TGraph   *gr;
+TEllipse *el;
+TBox     *bx;
+TPave    *pv;
+TDiamond *di;
+TCrown   *cr;
```
[the remark this answers is not in the document]
Comment (couet) [reply tutorials/graphics/inside.C:16]: Because they are used by the inside() and InPoint()

File: graf2d/graf/inc/TCrown.h (hunk 1 above)
Comment (vepadulano) [inline graf2d/graf/inc/TCrown.h:28]: Extra whitespace

File: graf2d/graf/inc/TCrown.h [hunk 3]
```diff
@@ -25,11 +25,12 @@ class TCrown : public TEllipse {
...
    Int_t   DistancetoPrimitive(Int_t px, Int_t py) override;
    virtual TCrown *DrawCrown(Double_t x1, Double_t y1, Double_t radin, Double_t radout,
                              Double_t  phimin=0, Double_t  phimax=360, Option_t *option="");
    void    ExecuteEvent(Int_t event, Int_t px, Int_t py) override;
+   Int_t   IsInside(Double_t x, Double_t y) const;
```
Comment (vepadulano) [inline graf2d/graf/inc/TCrown.h:33]: Missing `override`?

File: graf2d/graf/inc/TEllipse.h [hunk 4]
```diff
@@ -45,6 +45,7 @@ class TEllipse : public TObject, public TAttLine, public TAttFill, public TAttBB
    void                 Draw(Option_t *option="") override;
    virtual TEllipse    *DrawEllipse(Double_t x1, Double_t y1, Double_t r1,Double_t r2,Double_t phimin, Double_t phimax,Double_t theta,Option_t *option="");
    void                 ExecuteEvent(Int_t event, Int_t px, Int_t py) override;
+   Int_t                IsInside(Double_t x, Double_t y) const;
```
Comment (vepadulano) [inline graf2d/graf/inc/TEllipse.h:48]: Missing `override`?

**Review by vepadulano — CHANGES_REQUESTED** [review]

**Comment by hansenjo** [comment]
Hi guys,
I suggest a rewrite of the `TEllipse::IsInside` method as follows:
```
Int_t TEllipse::IsInside(Double_t x, Double_t y) const
{
  x -= fX1;
  y -= fY1;
  Double_t th = fTheta*TMath::DegToRad();
  Double_t st = TMath::Sin(th);
  Double_t ct = TMath::Cos(th);
  Double_t xx =  ct*x + st*y;
  Double_t yy = -st*x + ct*y;

  if (TMath::Abs(xx) > fR1 || TMath::Abs(yy) > fR2)
    return 0;
  Double_t xn = xx/fR1;
  Double_t yn = yy/fR2;
  if (xn*xn + yn*yn > 1.)
    return 0;
  if (fPhimax - fPhimin >= 360.)
    return 1;
  Double_t phimin = std::fmod(fPhimin, 360.);
  Double_t phimax = std::fmod(fPhimax, 360.);
  Double_t phi = TMath::RadToDeg()*(TMath::Pi() + TMath::ATan2(-yy*fR1/fR2, -xx));
  if (phi < phimin || phi > phimax)
    return 0;

  return 1;
}
```
Rationale:
- The algorithm in the PR is mathematically incorrect. For points close to the edge, it may fail to report a point as being inside. (My middle school math teacher would have failed this solution.)
- The code in the PR is inefficient: 404 calls to trig functions vs. at most 3 in the solution above.

You may want to add extra consistency checking (`fPhimin` < `fPhimax`, divisors != 0.0 etc.) in case the object was constructed with unreasonable parameters. Then again, such checks probably should go in the `TEllipse` constructor.

**Pull request closed without merging**